app-android/src/main/java/com/ustadmobile/port/android/view/binding/DatePickerBindingAdapter2.kt

```diff
@@ -0,0 +1,277 @@
+                            context,
+                            persianPickerDate.getPersianYear().toString() + "/" +
+                                    persianPickerDate.getPersianMonth() + "/" +
+                                        persianPickerDate.getPersianDay() + " timestamp: " + persianPickerDate.timestamp
+                                        + " converted : " + UMCalendarUtil.getPrettyDateFromLong(persianPickerDate.timestamp, null) ,
+                            Toast.LENGTH_SHORT
+                        ).show()
+
+                        adapterCalendar.timeInMillis = persianPickerDate.timestamp
+
+                        val persianDate = PersianDateImpl()
+                        persianDate.setDate(adapterCalendar.timeInMillis)
+
+                        text = persianPickerDate.getPersianYear().toString() + "/" +
+                                persianPickerDate.getPersianMonth() + "/" +
+                                persianPickerDate.getPersianDay()
+
+                        inverseBindingListener.onChange()
+                    }
+
+                    override fun onDismissed() {}
+                })
+
+            picker.show()
+
+        }else {
+
+            val builder = AlertDialog.Builder(context)
+                .setView(dialogView)
+
+            val picker = dialogView.findViewById<DatePicker>(R.id.date_picker)
+            picker.init(
+                defaultCal.get(Calendar.YEAR), defaultCal.get(Calendar.MONTH), defaultCal.get(Calendar.DAY_OF_MONTH),
+                null
+            )
+
+            builder.setPositiveButton(R.string.ok) { dialog, _ ->
+                cal[Calendar.DAY_OF_MONTH] = picker.dayOfMonth
+                cal[Calendar.MONTH] = picker.month
+                cal[Calendar.YEAR] = picker.year
+                text = adapterDateFormat.format(Date(adapterCalendar.timeInMillis))
+                inverseBindingListener.onChange()
+            }
+            builder.show()
+        }
+    }
+}
+
+@BindingAdapter("visibleIfDateSet")
+fun View.setVisibilityIfSetDate(date: Long){
+    visibility = if(date == 0L || date == Long.MAX_VALUE) View.GONE else View.VISIBLE
+}
+
+
+@BindingAdapter("relativeTime")
+fun TextView.setDateWithRelativeTime(date: Long){
+    text = DateUtils.getRelativeTimeSpanString(date)
+}
+
+@BindingAdapter("dateUseSpinners")
+fun TextView.setDateUseSpinners(dateUseSpinners: Boolean) {
+    setTag(R.id.tag_dateusespinner, dateUseSpinners)
+}
+
+@SuppressLint("SetTextI18n")
+@BindingAdapter(value = ["dateTimeMillisFrom", "dateTimeMillisTo"])
+fun TextView.setTextFromToDateTimeMillis(textFromDateLong: Long, textToDateLong: Long) {
+    val impl : UstadMobileSystemImpl = (context.applicationContext as DIAware
+            ).di.direct.instance()
+    val localeString = impl.getLocale(context)
+    val persianDateFrom = PersianDateImpl()
+    persianDateFrom.setDate(textFromDateLong)
+
+    val persianDateTo = PersianDateImpl()
```

Comment:
Let's move initializing these formatters into the if statement. There is no need to init the Persian calendar when it is not in use (and vice versa)